### PR TITLE
Make `refresh_top_level()` more like `visit_file()` in the third pass

### LINKED ISSUE #6032: Two crashes in fine grained mode with invalid types inside synthetic constructs
First traceback:
```
  File "mypy/dmypy_server.py", line 215, in serve
  File "mypy/dmypy_server.py", line 252, in run_command
  File "mypy/dmypy_server.py", line 300, in cmd_check
  File "mypy/dmypy_server.py", line 335, in check
  File "mypy/dmypy_server.py", line 382, in initialize_fine_grained
  File "mypy/server/update.py", line 235, in update
  File "mypy/server/update.py", line 287, in update_one
  File "mypy/server/update.py", line 342, in update_module
  File "mypy/server/update.py", line 549, in update_module_isolated
  File "mypy/server/update.py", line 715, in replace_modules_with_new_variants
  File "mypy/server/astmerge.py", line 88, in merge_asts
  File "mypy/server/astmerge.py", line 133, in replace_nodes_in_ast
  File "mypy/nodes.py", line 257, in accept
  File "mypy/server/astmerge.py", line 151, in visit_mypy_file__NodeVisitor_glue
  File "mypy/server/astmerge.py", line 154, in visit_mypy_file
  File "mypy/traverser.py", line 33, in visit_mypy_file
  File "mypy/nodes.py", line 968, in accept
  File "mypy/server/astmerge.py", line 195, in visit_assignment_stmt__StatementVisitor_glue
  File "mypy/server/astmerge.py", line 197, in visit_assignment_stmt
  File "mypy/traverser.py", line 79, in visit_assignment_stmt
  File "mypy/nodes.py", line 1421, in accept
  File "mypy/server/astmerge.py", line 231, in visit_call_expr__ExpressionVisitor_glue
  File "mypy/server/astmerge.py", line 232, in visit_call_expr
  File "mypy/traverser.py", line 166, in visit_call_expr
  File "mypy/nodes.py", line 2062, in accept
  File "mypy/server/astmerge.py", line 236, in visit_newtype_expr__ExpressionVisitor_glue
  File "mypy/server/astmerge.py", line 239, in visit_newtype_expr
  File "mypy/server/astmerge.py", line 318, in process_synthetic_type_info
  File "mypy/nodes.py", line 610, in accept__Node_glue
  File "mypy/nodes.py", line 611, in accept
  File "mypy/server/astmerge.py", line 160, in visit_func_def__StatementVisitor_glue
  File "mypy/server/astmerge.py", line 162, in visit_func_def
  File "mypy/server/astmerge.py", line 188, in process_base_func
  File "mypy/server/astmerge.py", line 292, in fixup_type
  File "mypy/types.py", line 838, in accept
  File "mypy/server/astmerge.py", line 346, in visit_callable_type__TypeVisitor_glue
  File "mypy/server/astmerge.py", line 348, in visit_callable_type
  File "mypy/types.py", line 541, in accept
  File "mypy/server/astmerge.py", line 335, in visit_instance__TypeVisitor_glue
  File "mypy/server/astmerge.py", line 338, in visit_instance
  File "mypy/types.py", line 1409, in accept
  File "mypy/server/astmerge.py", line 417, in visit_union_type__TypeVisitor_glue
  File "mypy/server/astmerge.py", line 419, in visit_union_type
  File "mypy/types.py", line 541, in accept
  File "mypy/server/astmerge.py", line 335, in visit_instance__TypeVisitor_glue
  File "mypy/server/astmerge.py", line 338, in visit_instance
  File "mypy/types.py", line 1597, in accept
  File "mypy/type_visitor.py", line 105, in visit_forwardref_type
RuntimeError: Internal error: unresolved forward reference
```
Second traceback:
```
File "mypy/dmypy_server.py", line 205, in serve
File "mypy/dmypy_server.py", line 242, in run_command
File "mypy/dmypy_server.py", line 290, in cmd_check
File "mypy/dmypy_server.py", line 325, in check
File "mypy/dmypy_server.py", line 372, in initialize_fine_grained
File "mypy/server/update.py", line 254, in update
File "mypy/server/update.py", line 767, in propagate_changes_using_dependencies
File "mypy/server/update.py", line 927, in reprocess_nodes
File "mypy/server/astdiff.py", line 154, in snapshot_symbol_table
File "mypy/server/astdiff.py", line 237, in snapshot_optional_type
File "mypy/server/astdiff.py", line 232, in snapshot_type
File "mypy/types.py", line 1409, in accept
File "mypy/server/astdiff.py", line 321, in visit_union_type__TypeVisitor_glue
File "mypy/server/astdiff.py", line 324, in visit_union_type
File "mypy/server/astdiff.py", line 232, in snapshot_type
File "mypy/types.py", line 1597, in accept
File "mypy/type_visitor.py", line 105, in visit_forwardref_type
RuntimeError: Internal error: unresolved forward reference
```

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #6032 (first part)

This is a very straightforward fix. I am actually surprised no one found this before (maybe because mypy daemon is not so widely used yet).

In future, it would be great to have a mechanism to avoid such mistakes, by somehow making `refresh_top_level()` to do all the special things `visit_file()` does. I don't have any ideas now, but it probably makes sense to consider this during semantic analyzer refactoring (another possible solution is to avoid any such special things, but this should be clearly documented).

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
Argh, forgot to run tests. Will fix them now.

**Comment by ilevkivskyi** [comment]
OK, now everything should be fine. I also updated PR description.

File: mypy/semanal_pass3.py [hunk 1]
```diff
@@ -471,12 +472,18 @@ def patch2() -> None:
...
+    def analyze_synthetic_info(self, info: TypeInfo) -> None:
         # Similar to above but for nodes with synthetic TypeInfos (NamedTuple and NewType).
         for name in info.names:
             sym = info.names[name]
             if isinstance(sym.node, (FuncDef, Decorator)):
+                # Since we are analyzing a synthetic type info, the methods there
+                # are not real independent targets, and should be processed when
+                # the enclosing synthetic type is processed.
+                old_recurse = self.recurse_into_functions
+                self.recurse_into_functions = True
```
Comment (JukkaL) [inline mypy/semanal_pass3.py]: If this is not needed to fix the test cases, could you move this to a separate PR?

Comment (ilevkivskyi) [reply mypy/semanal_pass3.py]: Hm, I think I can split it by moving one of the tests added to the second PR.
